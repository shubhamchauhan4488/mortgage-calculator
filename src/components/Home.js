import React, { useState } from "react";
import { MortgageForm } from "./MortgageForm";

export const Home = () => {
  const [mortgageAmt, setMortgageAmt] = useState(50000);
  const [rate, setRate] = useState(5);
  const [amortizationPeriod, setAmortizationPeriod] = useState(25);
  const [paymentFrequency, setPaymentFrequency] = useState(12);
  const [term, setTerm] = useState(5);
  const [errors, setErrors] = useState({
    mortgageAmt: "",
    rate: "",
    amortizationPeriod: "",
    paymentFrequency: "",
    term: "",
  });

  const handleInputChange = (e) => {
    setErrors({
      ...errors,
      [e.target.name]: "",
    });
    switch (e.target.name) {
      case "mortgageAmt": {
        setMortgageAmt(e.target.value);
        break;
      }
      case "rate": {
        setRate(e.target.value);
        break;
      }
      case "amortizationPeriod": {
        setAmortizationPeriod(e.target.value);
        break;
      }
      case "paymentFrequency": {
        setPaymentFrequency(e.target.value);
        break;
      }
      case "term": {
        setTerm(e.target.value);
        break;
      }
      default:
        return;
    }
  };

  return (
    <div className="main-content-container">
      <MortgageForm
        mortgageAmt={mortgageAmt}
        rate={rate}
        amortizationPeriod={amortizationPeriod}
        paymentFrequency={paymentFrequency}
        term={term}
        handleInputChange={handleInputChange}
        errors={errors}
      />
    </div>
  );
};
