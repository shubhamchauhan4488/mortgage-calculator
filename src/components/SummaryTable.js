import React from "react";

export const SummaryTable = ({ summaryTableData }) => {

  return (
    <div className="container-fluid">
      <h2 className="summary-heading">Summary</h2>
      <div className="row justify-content-center align-items-center mb-3 mt-3">
        <div className="col-12 col-sm-10">
          <table
            id="summary-table"
            data-testid="summary-table"
            className="table table-striped"
          >
            <thead className="thead-dark">
              <tr>
                <th scope="col">Category</th>
                <th scope="col">Term Figures</th>
                <th scope="col">Amortization Figures</th>
              </tr>
            </thead>
            <tbody>
              {summaryTableData.map((item, index) => (
                <tr key={index}>
                  <th scope="row">{item.category}</th>
                  <td id={`term-${index}`} data-testid={`term-${index}`}>
                    {item.term}
                  </td>
                  <td
                    id={`amortization-${index}`}
                    data-testid={`amortization-${index}`}
                  >
                    {item.amortization}
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  )
}

