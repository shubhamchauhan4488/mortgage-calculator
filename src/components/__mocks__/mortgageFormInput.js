export const mockMortgageFormProps = {
  mortgageAmt: '100000',
  rate: '5',
  amortizationPeriod: '25',
  paymentFrequency: '12',
  term: '5',
  handleInputChange: jest.fn(),
  handleInputBlur: jest.fn(),
  handleSubmit: jest.fn(),
  errors: {}
}