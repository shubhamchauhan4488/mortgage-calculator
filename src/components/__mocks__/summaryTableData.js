export const mockSummaryTableData = [
  {
    amortization: 600,
    category: "Number of Payments",
    term: 120
  },
  {
    amortization: "319.73",
    category: "Mortgage Payment",
    term: "319.73"
  },
  {
    amortization: "100000.00",
    category: "Principal Payments,",
    term: "10019.40"
  },
  {
    amortization: "91838.89",
    category: "Interest Payments",
    term: "28348.20",
  },
  {
    amortization: "191838.89",
    category: "Total Cost",
    term: "38367.78"
  }
]