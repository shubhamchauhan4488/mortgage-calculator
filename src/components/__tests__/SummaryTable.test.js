import React from 'react';
import { fireEvent, cleanup, render } from '@testing-library/react';
import { SummaryTable } from '../SummaryTable';
import { mockSummaryTableData } from '../__mocks__/summaryTableData';
import { Home } from '../Home';

describe('Summary Table', () => {
  afterEach(cleanup);

  it('renders correctly', () => {
    const { asFragment } = render(<SummaryTable summaryTableData={mockSummaryTableData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Summary table correct data should be rendered', () => {
    const { queryByTestId, getAllByText, getByText } = render(<Home />);
    const mortgageAmt = queryByTestId('mortgage-amt');
    const rateInput = queryByTestId('rate');
    const amortizatoinPeriodInput = queryByTestId('amortization-period');
    const paymentFrequencyInput = queryByTestId('payment-frequency');
    const termInput = queryByTestId('term');
    const calculateBtn = queryByTestId('calculate-btn');

    fireEvent.change(mortgageAmt, { target: { value: '100000' } });
    fireEvent.change(rateInput, { target: { value: '6' } });
    fireEvent.change(amortizatoinPeriodInput, { target: { value: '25' } });
    fireEvent.change(paymentFrequencyInput, { target: { value: '12' } });
    fireEvent.change(termInput, { target: { value: '5' } });
    fireEvent.click(calculateBtn);

    /** Term column figures */
    expect(getAllByText('639.81').length).toEqual(2)
    expect(getByText(/10163.47/)).toBeTruthy();
    expect(getByText(/28225.12/)).toBeTruthy();
    expect(getByText(/38388.40/)).toBeTruthy();

    /** Amortization column figures */
    expect(getByText(/100000.00/)).toBeTruthy();
    expect(getByText('91941.99')).toBeTruthy();
    expect(getByText('191941.99')).toBeTruthy();
  });

  it('Summary table should not be rendered in case of any error', () => {
    const { queryByTestId } = render(<Home />);
    const mortgageAmt = queryByTestId('mortgage-amt');
    const rateInput = queryByTestId('rate');
    const amortizatoinPeriodInput = queryByTestId('amortization-period');
    const paymentFrequencyInput = queryByTestId('payment-frequency');
    const termInput = queryByTestId('term');
    const calculateBtn = queryByTestId('calculate-btn');

    fireEvent.change(mortgageAmt, { target: { value: '1000000' } });
    fireEvent.change(rateInput, { target: { value: '105' } });
    fireEvent.change(amortizatoinPeriodInput, { target: { value: '25' } });
    fireEvent.change(paymentFrequencyInput, { target: { value: '12' } });
    fireEvent.change(termInput, { target: { value: '5' } });
    fireEvent.click(calculateBtn);
    expect(queryByTestId('summary-table')).not.toBeTruthy();
  });
});