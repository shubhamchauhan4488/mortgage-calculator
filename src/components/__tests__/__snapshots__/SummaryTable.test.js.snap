// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Table renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container-fluid"
  >
    <h2
      class="summary-heading"
    >
      Summary
    </h2>
    <div
      class="row justify-content-center align-items-center mb-3 mt-3"
    >
      <div
        class="col-12 col-sm-10"
      >
        <table
          class="table table-striped"
          data-testid="summary-table"
          id="summary-table"
        >
          <thead
            class="thead-dark"
          >
            <tr>
              <th
                scope="col"
              >
                Category
              </th>
              <th
                scope="col"
              >
                Term Figures
              </th>
              <th
                scope="col"
              >
                Amortization Figures
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                scope="row"
              >
                Number of Payments
              </th>
              <td
                data-testid="term-0"
                id="term-0"
              >
                120
              </td>
              <td
                data-testid="amortization-0"
                id="amortization-0"
              >
                600
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                Mortgage Payment
              </th>
              <td
                data-testid="term-1"
                id="term-1"
              >
                319.73
              </td>
              <td
                data-testid="amortization-1"
                id="amortization-1"
              >
                319.73
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                Principal Payments,
              </th>
              <td
                data-testid="term-2"
                id="term-2"
              >
                10019.40
              </td>
              <td
                data-testid="amortization-2"
                id="amortization-2"
              >
                100000.00
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                Interest Payments
              </th>
              <td
                data-testid="term-3"
                id="term-3"
              >
                28348.20
              </td>
              <td
                data-testid="amortization-3"
                id="amortization-3"
              >
                91838.89
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                Total Cost
              </th>
              <td
                data-testid="term-4"
                id="term-4"
              >
                38367.78
              </td>
              <td
                data-testid="amortization-4"
                id="amortization-4"
              >
                191838.89
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
